engine/src/main/java/org/terasology/i18n/TranslationSystemImpl.java

```diff
@@ -0,0 +1,125 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.terasology.i18n;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.Consumer;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terasology.assets.ResourceUrn;
+import org.terasology.assets.management.AssetManager;
+import org.terasology.config.Config;
+import org.terasology.config.SystemConfig;
+import org.terasology.context.Context;
+import org.terasology.engine.SimpleUri;
+import org.terasology.engine.Uri;
+import org.terasology.i18n.assets.Translation;
+
+/**
+ * A translation system that uses {@link Translation} data assets to
+ * perform the lookup.
+ */
+public class TranslationSystemImpl implements TranslationSystem {
+
+    private static final Logger logger = LoggerFactory.getLogger(TranslationSystemImpl.class);
+
+    private final List<Consumer<TranslationProject>> changeListeners = new CopyOnWriteArrayList<>();
+    private final Map<Uri, TranslationProject> projects = new HashMap<>();
+
+    private final SystemConfig config;
+
+    /**
+     * @param context the context to use
+     */
+    public TranslationSystemImpl(Context context) {
+
+        config = context.get(Config.class).getSystem();
+        AssetManager assetManager = context.get(AssetManager.class);
+
+        Set<ResourceUrn> urns = assetManager.getAvailableAssets(Translation.class);
+        for (ResourceUrn urn : urns) {
+            Optional<Translation> asset = assetManager.getAsset(urn, Translation.class);
+            if (asset.isPresent()) {
+                Translation trans = asset.get();
+                Uri uri = trans.getProjectUri();
+                if (uri.isValid()) {
+                    TranslationProject proj = projects.computeIfAbsent(uri, e -> new StandardTranslationProject());
+                    proj.add(trans);
+                    trans.subscribe(this::onAssetChanged);
+                    logger.info("Discovered " + trans);
+                } else {
+                    logger.warn("Ignoring invalid project uri: {}", uri);
+                }
+            }
+        }
+    }
+
+    @Override
+    public TranslationProject getProject(Uri name) {
+        return projects.get(name);
+    }
+
+    @Override
+    public String translate(String id) {
+        return translate(id, config.getLocale());
+    }
+
+    @Override
+    public String translate(String id, Locale otherLocale) {
+        int splitPoint = id.indexOf('#');
+        if (splitPoint > 0) {
+            String projName = id.substring(0, splitPoint);
```

Comment:
You could also load it into an AssetUri, which naturally splits out the different parts?